Initial abstractions

Jevvry opened the pull request.

Jevvry committed:
Add initial models

TGladysheva commented on Vostok.ServiceDiscovery.Telemetry/Event/ServiceDiscoveryEventKind.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        AddToBlackList = 0,
+        RemoveFromBlackList = 1,
+        ReplicaStart = 2,
```
Я бы ReplicaStart и ReplicaStop в начало списка перенесла

TGladysheva commented on Vostok.ServiceDiscovery.Telemetry/ServiceDiscoveryContext.cs:
```diff
@@ -0,0 +1,19 @@
+    [PublicAPI]
+    public static class ServiceDiscoveryContext
+    {
+        private static readonly AsyncLocal<IServiceDiscoveryEventsBuilder> Container = new AsyncLocal<IServiceDiscoveryEventsBuilder>();
```
Есть ощущение, что весь билдер тут не нужен. 
Как будто бы мы этот контекст будем использовать из SDManagerExtersions. А там у нас не будет EventKind'а и свойств кто-когда-почему добавил реплику в блеклист.

Более того, заполнять часть этих свойств мы будем непосредственно из мониторинга SD, т.к. в методы SDManager его не прокинуть. Это, кажется, не ответственность мониторинга, знать, как создавать SDEvent'ы.

Можно создать класс попроще, с ограниченным набором свойств

kungasc commented on Vostok.ServiceDiscovery.Telemetry/Event/ServiceDiscoveryEvent.cs:
```diff
@@ -0,0 +1,41 @@
+            [NotNull] string application,
+            [NotNull] string replica,
+            ServiceDiscoveryEventKind eventKind,
+            [NotNull] string environment = "default",
```
вместо опциональных параметров надо использовать сетер

public string Environment { get; set; } = "default"

но вообще вроде бы никакого дефолта "default" в либах SD нету, и этой либе он не нужен

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventBuilder/ServiceDiscoveryEventsBuilder.cs:
```diff
@@ -0,0 +1,29 @@
+namespace Vostok.ServiceDiscovery.Telemetry.EventBuilder
+{
+    [PublicAPI]
+    public class ServiceDiscoveryEventsBuilder: IServiceDiscoveryEventsBuilder
```
а зачем этот билдер, у вас правда эвент по кусочкам собирать нужно, а не просто создать и записать?

аналогично с ServiceDiscoveryContext

TGladysheva replied:
>а зачем этот билдер, у вас правда эвент по кусочкам собирать нужно, а не просто создать и записать?


Да, правда

kungasc commented on Vostok.ServiceDiscovery.Telemetry/ServiceDiscoveryContext.cs:
```diff
@@ -0,0 +1,19 @@
+namespace Vostok.ServiceDiscovery.Telemetry
+{
+    [PublicAPI]
+    public static class ServiceDiscoveryContext
```
не знаю, что делает этот класс (даже доки нету), но предположу что название не соответствует действительности

если он только про отправку эвентов, а не про что-то общее из мира ServiceDiscovery

TGladysheva replied:
Да, мы обсудили, что в его названии должно быть слово Event или Operation

Jevvry committed:
Add service discovery event description

Jevvry committed:
Refactor service discovery event

Jevvry committed:
Merge branch 'master' into initial_models

Jevvry committed:
Refactor event creation

Jevvry committed:
Add dependencies wellknown property

Jevvry committed:
Store replica uri as string

Jevvry committed:
Add event sender extensions

Jevvry committed:
Add service discovery event description implementation

Jevvry committed:
Add events builder tests

TGladysheva commented on Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventDescription.cs:
```diff
@@ -0,0 +1,52 @@
+            throw new NotImplementedException();
+
+        [NotNull]
+        public ServiceDiscoveryEventDescription AddProperties(params (string key, string value)[] properties) =>
```
Я бы сделала метод, который просто key value принимает, а не массив.

TGladysheva replied:
Все равно у нас на разных уровнях вложенности это может заполняться.

TGladysheva commented on Vostok.ServiceDiscovery.Telemetry/ServiceDiscoveryEventKeys.cs:
```diff
@@ -0,0 +1,13 @@
+    [PublicAPI]
+    public static class ServiceDiscoveryEventKeys
+    {
+        public const string EventCreatorId = "userId";
```
Слово Event можно убрать из названия полей, оно есть в названии класса

TGladysheva commented on Vostok.ServiceDiscovery.Telemetry/Event/ServiceDiscoveryEvent.cs:
```diff
@@ -0,0 +1,41 @@
+            [NotNull] string application,
+            [NotNull] string replica,
+            [NotNull] string environment,
+            ServiceDiscoveryEventKind eventKind,
```
Может eventKind первым аргументом?

Jevvry committed:
Change event constructor

Jevvry committed:
Rename well-known keys

Jevvry committed:
Refactor event description

Jevvry committed:
Change event constructor

Jevvry committed:
Rename well-known keys

Jevvry committed:
Refactor service discovery event description

Jevvry committed:
Rename description extensions

Jevvry committed:
Make scoped event description

Jevvry committed:
Merge branch 'initial_models' into models_implementation

# Conflicts:
#	Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventBuilder.cs
#	Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventDescription.cs
#	Vostok.ServiceDiscovery.Telemetry/Extensions/ServiceDiscoveryEventDescriptionExtensions.cs

Jevvry committed:
Apply merge

Jevvry committed:
Add event sender extensions

Jevvry committed:
Add explicit use to flowing context

Jevvry committed:
Add service discovery event description context tests

TGladysheva commented on Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventDescription.cs:
```diff
@@ -1,50 +1,61 @@
-        private readonly List<string> replicas = new List<string>();
+        private readonly Dictionary<string, string> properties;
+        private readonly List<string> replicas;
+        private IDisposable descriptionScope;
```
ISDEventDescription лучше, а не IDisposable

Jevvry replied:
Scope задается вот в этой [части](https://github.com/vostok/servicediscovery.telemetry/blob/initial_models/Vostok.ServiceDiscovery.Telemetry/ServiceDiscoveryEventDescriptionContext.cs#L33) и он как раз хранит в себе логику замены на прежний `IServiceDiscoveryEventDescription` при диспоузе текущего. Нам ведь не нужны методы `IServiceDiscoveryEventDescription` для скоупа, и наверное не стоит пытаться обернуть логику `FlowingContext.Globals.Use` в наш `IServiceDiscoveryEventDescription`.

kungasc replied:
тут всё ок, IDisposable контекста

Jevvry committed:
Merge branch 'master' into initial_models

kungasc commented on Vostok.ServiceDiscovery.Telemetry/Event/ServiceDiscoveryEventKind.cs:
```diff
@@ -0,0 +1,13 @@
+    [PublicAPI]
+    public enum ServiceDiscoveryEventKind
+    {
+        ReplicaStart = 0,
```
ReplicaStart vs AddToBlackList

вроде как все эвенты про реплику, поэтому хватит и просто Start/Stop

либо везде с Replica (ReplicaAddedToBlacklist)

Jevvry replied:
Лучше везде реплику дописать

kungasc committed:
Merge remote-tracking branch 'origin/models_implementation' into initial_models

kungasc committed:
fix event models nitpicks

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventSender/IServiceDiscoveryEventSender.cs:
```diff
@@ -0,0 +1,11 @@
+    [PublicAPI]
+    public interface IServiceDiscoveryEventSender
+    {
+        void Send([NotNull] ServiceDiscoveryEvent serviceDiscoveryEvent);
```
```suggestion
        void Send([NotNull] ServiceDiscoveryEvent event);
```

kungasc commented on Vostok.ServiceDiscovery.Telemetry/Extensions/IServiceDiscoveryEventDescriptionExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+namespace Vostok.ServiceDiscovery.Telemetry.Extensions
+{
+    [PublicAPI]
+    public static class IServiceDiscoveryEventDescriptionExtensions
```
экстеншены лучше держать в той же папке, что и соответствующая модель

kungasc commented on Vostok.ServiceDiscovery.Telemetry/Extensions/IServiceDiscoveryEventDescriptionExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+
+        [NotNull]
+        public static IServiceDiscoveryEventDescription SetUserId(this IServiceDiscoveryEventDescription serviceDiscoveryEventDescription, [NotNull] string userId) =>
+            serviceDiscoveryEventDescription.AddProperty(ServiceDiscoveryWellKnownProperties.CreatorId, userId);
```
поля разъехались как будто (creatorId vs userId)

лучше везде одинаково назвать

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventDescription.cs:
```diff
@@ -0,0 +1,81 @@
+
+namespace Vostok.ServiceDiscovery.Telemetry.EventDescription
+{
+    internal class ServiceDiscoveryEventDescription : IServiceDiscoveryEventDescription
```
по сути это и есть ServiceDiscoveryEventsBuilder

не понятно, зачем два отдельных класса, лучше этот переименовать в билдер и сделать ему поле CurrentEvents (см ISpanBuilder)

Jevvry replied:
Изначально действительно было что-то похожее. Контекст отдавал [`IServiceDiscoveryEventsBuilder`](https://github.com/vostok/servicediscovery.telemetry/blob/5e9130fca29b713c8c026ca32a433fc51d9450b6/Vostok.ServiceDiscovery.Telemetry/EventBuilder/IServiceDiscoveryEventsBuilder.cs). Но выше есть [комментарий ](https://github.com/vostok/servicediscovery.telemetry/pull/1#discussion_r732774203) и это решили изменить

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventDescription/IServiceDiscoveryEventDescription.cs:
```diff
@@ -0,0 +1,40 @@
+        string Environment { get; }
+
+        [NotNull]
+        IReadOnlyList<string> Replicas { get; }
```
а зачем столько публичных полей торчит кусочных?

как будто ни одно в итоге пользователю и не нужно

Jevvry replied:
Так получилось как раз из-за того что `ServiceDiscoveryEventsBuilder` отдельный класс и ему нужен доступ к этим свойствам.

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventDescription/IServiceDiscoveryEventDescription.cs:
```diff
@@ -0,0 +1,40 @@
+        IServiceDiscoveryEventDescription SetEventKind(ServiceDiscoveryEventKind eventKind);
+
+        [NotNull]
+        IServiceDiscoveryEventDescription AddProperty([NotNull] string key, [NotNull] string value);
```
```suggestion
        IServiceDiscoveryEventDescription SetProperty([NotNull] string key, [NotNull] string value);
```

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventBuilder.cs:
```diff
@@ -0,0 +1,27 @@
+                serviceDiscoveryEventDescription.Environment,
+                serviceDiscoveryEventDescription.Application,
+                replica,
+                DateTimeOffset.UtcNow,
```
```suggestion
                DateTimeOffset.Now,
```

kungasc replied:
иначе нет смысла от оффсета

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventBuilder.cs:
```diff
@@ -0,0 +1,27 @@
+                serviceDiscoveryEventDescription.Environment,
+                serviceDiscoveryEventDescription.Application,
+                replica,
+                DateTimeOffset.UtcNow,
```
логичней кажется таймстемп заполнять при создании эвента, а в билде

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventDescription.cs:
```diff
@@ -0,0 +1,81 @@
+            properties = new Dictionary<string, string>();
+        }
+
+        public ServiceDiscoveryEventDescription(IServiceDiscoveryEventDescription eventDescription)
```
```suggestion
        public ServiceDiscoveryEventDescription Clone()
```

kungasc commented:
А ещё не хватает документаций на всех публичных методах/модельках.

И там же примеров использований (см ISpanBuilder).

kungasc commented on Vostok.ServiceDiscovery.Telemetry/ServiceDiscoveryEventDescriptionContext.cs:
```diff
@@ -0,0 +1,38 @@
+namespace Vostok.ServiceDiscovery.Telemetry
+{
+    [PublicAPI]
+    public static class ServiceDiscoveryEventDescriptionContext
```
```suggestion
    public static class ServiceDiscoveryEventsContext
```

Jevvry committed:
Rename ServiceDiscoveryEventDescriptionContext

Co-authored-by: kungurtsev <kungurtsev@skbkontur.ru>

Jevvry committed:
Update Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventDescription.cs

Co-authored-by: kungurtsev <kungurtsev@skbkontur.ru>

Jevvry committed:
Update Vostok.ServiceDiscovery.Telemetry/EventDescription/ServiceDiscoveryEventBuilder.cs

Co-authored-by: kungurtsev <kungurtsev@skbkontur.ru>

Jevvry committed:
Update Vostok.ServiceDiscovery.Telemetry/EventDescription/IServiceDiscoveryEventDescription.cs

Co-authored-by: kungurtsev <kungurtsev@skbkontur.ru>

Jevvry committed:
Update Vostok.ServiceDiscovery.Telemetry/EventSender/IServiceDiscoveryEventSender.cs

Co-authored-by: kungurtsev <kungurtsev@skbkontur.ru>

Jevvry committed:
Refactor naming

Jevvry committed:
Move extensions to model folder

Jevvry committed:
Add service discovery events builder

Jevvry committed:
Add service discovery event context

Jevvry committed:
Add service discovery event context token

Jevvry committed:
Tests:
    - add service discovery event context token tests
    - add service discovery events builder tests

Jevvry committed:
Public API documentation

Jevvry committed:
Merge branch 'master' into initial_models

# Conflicts:
#	Vostok.ServiceDiscovery.Telemetry.Tests/Vostok.ServiceDiscovery.Telemetry.Tests.csproj

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventContext/IServiceDiscoveryEventContext.cs:
```diff
@@ -0,0 +1,33 @@
+    /// See <see cref="CurrentEvents"/>, <see cref="Send"/>, <see cref="TrySendFromContext"/> for details. 
+    /// </summary>
+    [PublicAPI]
+    public interface IServiceDiscoveryEventContext
```
```suggestion
    public interface IServiceDiscoveryEventsContext
```

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventContext/IServiceDiscoveryEventContext.cs:
```diff
@@ -0,0 +1,33 @@
+using Vostok.ServiceDiscovery.Telemetry.Event;
+using Vostok.ServiceDiscovery.Telemetry.EventsBuilder;
+
+namespace Vostok.ServiceDiscovery.Telemetry.EventContext
```
```suggestion
namespace Vostok.ServiceDiscovery.Telemetry.EventsContext
```

kungasc replied:
А вообще основные интерфейсы/имплементации лучше вынести в корень.

(ServiceDiscoveryEventContext, IServiceDiscoveryEventContext, ServiceDiscoveryEventContextConfig, ServiceDiscoveryEventContextToken)

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventContext/IServiceDiscoveryEventContext.cs:
```diff
@@ -0,0 +1,33 @@
+        /// <para>Empty if environment, application and replicas is not set.</para>
+        /// </summary>
+        [NotNull]
+        IReadOnlyList<ServiceDiscoveryEvent> CurrentEvents { get; }
```
```suggestion
        IEnumerable<ServiceDiscoveryEvent> CurrentEvents { get; }
```

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventContext/IServiceDiscoveryEventContext.cs:
```diff
@@ -0,0 +1,33 @@
+        void Send(ServiceDiscoveryEvent @event);
+        
+        /// <summary>
+        /// Sends events from <see cref="CurrentEvents"/> using <see cref="ServiceDiscoveryEventContextConfig.Sender"/>.
```
And returns?

kungasc replied:
И как-то не понятно, как использовать этот bool.

Может лучше просто SendFromContext, и писать ошибки в ErrorCallback если чего-то не хватило.

Jevvry replied:
И кажется, что пустая коллекция ивентов не является ошибкой.  А ошибки при отправке считается, что sender должен сам обрабатывать,  поэтому кажется, что и ErrorCallback пока не нужен

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventContext/ServiceDiscoveryEventContext.cs:
```diff
@@ -0,0 +1,41 @@
+
+        public IReadOnlyList<ServiceDiscoveryEvent> CurrentEvents
+        {
+            get => FlowingContext.Globals.Get<ServiceDiscoveryEventsBuilder>()?.BuildEvents() ?? new List<ServiceDiscoveryEvent>();
```
не надо так везде сорить) 

use Array.Empty/Enumerable.Emtpy

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventContext/ServiceDiscoveryEventContextToken.cs:
```diff
@@ -0,0 +1,47 @@
+        /// </summary>
+        public void Dispose()
+        {
+            scope?.Dispose();
```
а как тут может быть scope = null

kungasc commented on Vostok.ServiceDiscovery.Telemetry/EventSender/DevNullServiceDiscoveryEventSender.cs:
```diff
@@ -0,0 +1,16 @@
+﻿using JetBrains.Annotations;
+using Vostok.ServiceDiscovery.Telemetry.Event;
+
+namespace Vostok.ServiceDiscovery.Telemetry.EventSender
```
тут возможно тоже должно быть EventsSender

Jevvry committed:
Update Vostok.ServiceDiscovery.Telemetry/EventContext/IServiceDiscoveryEventContext.cs

Co-authored-by: kungurtsev <kungurtsev@skbkontur.ru>

Jevvry committed:
Update Vostok.ServiceDiscovery.Telemetry/EventContext/IServiceDiscoveryEventContext.cs

Co-authored-by: kungurtsev <kungurtsev@skbkontur.ru>

Jevvry committed:
Update Vostok.ServiceDiscovery.Telemetry/EventContext/IServiceDiscoveryEventContext.cs

Co-authored-by: kungurtsev <kungurtsev@skbkontur.ru>

Jevvry committed:
Refactor main implementations

Jevvry committed:
Apply rename

Jevvry committed:
Small builder fix

Jevvry committed:
Fix events context token tests

Jevvry committed:
Use $(ReferencesFramework)

kungasc committed:
nitpicks 💅

Jevvry committed:
Update changelog

Jevvry committed:
Bumped version to 0.1.0
